File: app/src/main/java/linusfessler/alarmtiles/shared/VibrationManager.java
package linusfessler.alarmtiles.shared;

import android.os.Build;
import android.os.VibrationEffect;
import android.os.Vibrator;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class VibrationManager {

    private static final int DEFAULT_DURATION = 250;
    private static final int DEFAULT_AMPLITUDE = -1;

    private final Vibrator vibrator;

    @Inject
    public VibrationManager(final Vibrator vibrator) {
        this.vibrator = vibrator;
    }

    public void vibrate() {
        vibrate(DEFAULT_DURATION);
    }

    public void vibrate(final long milliseconds) {
        vibrate(milliseconds, DEFAULT_AMPLITUDE);
    }

    @SuppressWarnings("squid:CallToDeprecatedMethod")
    public void vibrate(final long milliseconds, final int amplitude) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(milliseconds, amplitude));
        } else {
            vibrator.vibrate(milliseconds);
        }
    }
}

File: app/src/main/java/linusfessler/alarmtiles/shared/VibrationManager.kt
package linusfessler.alarmtiles.shared

import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class VibrationManager @Inject constructor(private val vibrator: Vibrator) {
    fun vibrate(milliseconds: Long = DEFAULT_DURATION, amplitude: Int = DEFAULT_AMPLITUDE) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(milliseconds, amplitude))
        } else {
            @Suppress("deprecation")
            vibrator.vibrate(milliseconds)
        }
    }

    fun vibrate(pattern: LongArray = DEFAULT_VIBRATION_PATTERN) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createWaveform(pattern, 0));
        } else {
            @Suppress("deprecation")
            vibrator.vibrate(pattern, 0);
        }
    }

    fun cancel() {
        vibrator.cancel()
    }

    companion object {
        private const val DEFAULT_DURATION = 250L
        private const val DEFAULT_AMPLITUDE = -1
        private val DEFAULT_VIBRATION_PATTERN = longArrayOf(500, 500)
    }
}